### ISSUE TITLE
vslib fail to start syncd due to various warnings/errors

### ISSUE THREAD

**Issue by sydlogan** [Azure/sonic-sairedis#584]
The following is seen when we try to start vslib based switch:

Mar 27 20:21:35.663077 sonic WARNING syncd#syncd: :- refresh_read_only: need to recalculate RO: SAI_SWITCH_ATTR_SUPPORTED_OBJECT_TYPE_LIST
Mar 27 20:21:35.663077 sonic ERR syncd#syncd: :- get: SAI_SWITCH_ATTR_SUPPORTED_OBJECT_TYPE_LIST read only not implemented on oid:0x2100000000

The following patch (see https://github.com/Azure/sonic-sairedis/pull/583) seems to address this issue.
```
diff --git a/vslib/src/SwitchStateBase.cpp b/vslib/src/SwitchStateBase.cpp
index 8ac4c64..4ef8e23 100644
--- a/vslib/src/SwitchStateBase.cpp
+++ b/vslib/src/SwitchStateBase.cpp
@@ -684,6 +684,12 @@ sai_status_t SwitchStateBase::set_switch_default_attributes()
 {
     SWSS_LOG_ENTER();
 
+    int32_t supported_obj_list[] = {
+                                SAI_OBJECT_TYPE_NULL,
+                                SAI_OBJECT_TYPE_PORT,
+                                SAI_OBJECT_TYPE_LAG
+                              };
+
     SWSS_LOG_INFO("create switch default attributes");
 
     sai_attribute_t attr;
@@ -725,6 +731,12 @@ sai_status_t SwitchStateBase::set_switch_default_attributes()
     attr.id = SAI_SWITCH_ATTR_WARM_RECOVER;
     attr.value.booldata = false;
 
+    CHECK_STATUS(set(SAI_OBJECT_TYPE_SWITCH, m_switch_id, &attr));
+
+    attr.id = SAI_SWITCH_ATTR_SUPPORTED_OBJECT_TYPE_LIST;
+    attr.value.s32list.count = sizeof(supported_obj_list);
+    attr.value.s32list.list = supported_obj_list;
+
     return set(SAI_OBJECT_TYPE_SWITCH, m_switch_id, &attr);
 }
 
@@ -1538,6 +1550,7 @@ sai_status_t SwitchStateBase::refresh_read_only(
             case SAI_SWITCH_ATTR_DEFAULT_VLAN_ID:
             case SAI_SWITCH_ATTR_DEFAULT_STP_INST_ID:
             case SAI_SWITCH_ATTR_DEFAULT_1Q_BRIDGE_ID:
+            case SAI_SWITCH_ATTR_SUPPORTED_OBJECT_TYPE_LIST:
                 return SAI_STATUS_SUCCESS;
 
             case SAI_SWITCH_ATTR_ACL_ENTRY_MINIMUM_PRIORITY:
```
The above patch simply is a copy of code that was present before bigrebase was merged.

However, additional failures to find attributes in vslib are seen. The following is just a small snippet from syslog:

Mar 28 00:00:54.984287 sonic NOTICE syncd#syncd: :- init_switch: init took 0.282957 sec
Mar 28 00:00:54.990609 sonic ERR syncd#syncd: :- get: SAI_SWITCH_ATTR_SWITCH_HARDWARE_INFO not implemented on oid:0x2100000000
Mar 28 00:00:54.994738 sonic WARNING syncd#syncd: :- saiGetHardwareInfo: failed to get switch hardware info: SAI_STATUS_NOT_IMPLEMENTED
Mar 28 00:00:54.994738 sonic NOTICE syncd#syncd: :- refresh_port_list: refreshed port list, current port number: 56, not counting cpu port
Mar 28 00:00:55.001997 sonic ERR syncd#syncd: :- get: SAI_QUEUE_ATTR_PARENT_SCHEDULER_NODE not implemented on oid:0x1500000000
Mar 28 00:00:55.001997 sonic ERR syncd#syncd: :- get: SAI_QUEUE_ATTR_WRED_PROFILE_ID not implemented on oid:0x1500000000
Mar 28 00:00:55.001997 sonic ERR syncd#syncd: :- get: SAI_QUEUE_ATTR_BUFFER_PROFILE_ID not implemented on oid:0x1500000000
Mar 28 00:00:55.001997 sonic ERR syncd#syncd: :- get: SAI_QUEUE_ATTR_SCHEDULER_PROFILE_ID not implemented on oid:0x1500000000
Mar 28 00:00:55.001997 sonic ERR syncd#syncd: :- get: SAI_QUEUE_ATTR_TAM_OBJECT not implemented on oid:0x1500000000
Mar 28 00:00:55.001997 sonic ERR syncd#syncd: :- get: SAI_QUEUE_ATTR_PARENT_SCHEDULER_NODE not implemented on oid:0x1500000001
Mar 28 00:00:55.001997 sonic ERR syncd#syncd: :- get: SAI_QUEUE_ATTR_WRED_PROFILE_ID not implemented on oid:0x1500000001
Mar 28 00:00:55.001997 sonic ERR syncd#syncd: :- get: SAI_QUEUE_ATTR_BUFFER_PROFILE_ID not implemented on oid:0x1500000001

The saiGetHardwareInfo appears to be expected (per comments in the code), the others appear to 
be enough for eventually swss and syncd to fail.

**Comment by kcudnik** [comment]
to support SAI_SWITCH_ATTR_SUPPORTED_OBJECT_TYPE_LIST first SAI pointer must be updated to actually include this change

**Comment by sydlogan** [comment]
Closing this issue. It is agreed we will need to move SAI to support additional switch devices/phys in the future. The root cause of this defect is that Broadcom is currently working with a newer SAI than upstream. We've worked thru this issue by patching sonic-sairedis in support for the missing attributes reported. Those changes will need to make their way upstream to support newer SAI when the SAI pointer is moved to support the SAI we are currently working with.
